renku/core/models/provenance/datasets.py

```diff
@@ -0,0 +1,515 @@
+        files = _convert_dataset_files(dataset.files, client, revision)
+
+        return Dataset(
+            name=dataset.name,
+            client=client,
+            creators=dataset.creators,
+            date_created=dataset.date_created,
+            date_deleted=None,
+            date_published=dataset.date_published,
+            derived_from=dataset.derived_from,
+            description=dataset.description,
+            files=files,
+            id=None,
+            identifier=dataset.identifier,
+            in_language=dataset.in_language,
+            keywords=dataset.keywords,
+            license=dataset.license,
+            original_identifier=dataset.original_identifier,
+            same_as=dataset.same_as,
+            tags=dataset.tags,
+            title=dataset.title,
+            url=dataset.url,
+            version=dataset.version,
+        )
+
+    @property
+    def client(self):
+        """Return client."""
+        return self._client
+
+    @client.setter
+    def client(self, client):
+        """Set client."""
+        self._update_client(client)
+
+    def is_deleted(self):
+        """Return true if dataset is deleted and should not be accessed."""
+        return self.date_deleted is not None
+
+    def find_files(self, paths):
+        """Return all paths that are in files container."""
+        files_paths = {str(self.client.path / f.path) for f in self.files}
+        return {p for p in paths if str(p) in files_paths}
+
+    def find_file(self, path, return_index=False):
+        """Find a file in files container using its relative path."""
+        for index, file_ in enumerate(self.files):
+            if str(file_.path) == str(path):
+                if return_index:
+                    return index
+                file_.client = self.client
+                return file_
+
+    def _set_identifier(self, new_identifier):
+        """Set identifier and update all related fields."""
+        self.identifier = new_identifier
+        self.id = generate_dataset_id(client=self.client, identifier=self.identifier)
+        self.url = self.id
+
+    def _update_client(self, client):
+        """Set new client and update related fields."""
+        self._client = client
+
+        self._set_identifier(self.identifier)
+
+        if self.derived_from:
+            host = get_host(self.client)
+            derived_from_id = self.derived_from._id
+            derived_from_url = self.derived_from.url.get("@id")
+            u = urlparse(derived_from_url)
+            derived_from_url = u._replace(netloc=host).geturl()
+            self.derived_from = Url(id=derived_from_id, url_id=derived_from_url)
+
+        for file_ in self.files:
+            file_.client = client
+
+    def update_from(self, dataset, client, revision, date):
+        """Update metadata from a new version of the dataset."""
+        self._set_identifier(dataset.identifier)
```

Comment:
Does this go against dataset immutability? I.e. using this you can change the properties of a dataset without generating a new ID?

Reply:
Right. Actually, when dataset has a non-existing identifier we insert it as a new dataset and this function won't get called. This is called only when identifier is the same as the current identifier, so, we don't need to set it again.